### PR TITLE
Upgrade mechanism (#23).

### PR DESCRIPTION
**Pull request by paulodamaso**
For #23: Created tests for `HtUpgradeWire` behavior.

### CODE REVIEW

**Comment by paulodamaso** [comment]
@llorllale Could yout take a look at this, looks like Zerocrat is not assigning this `REV` task to anyone.

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/yegor256/cactoos-http/issues/53#issuecomment-389216670))

@llorllale This is what I know about this job, as in [§32](http://www.zerocracy.com/policy.html#32):

  * The job [#53](https://github.com/yegor256/cactoos-http/issues/53) is [in scope](http://www.zerocracy.com/policy.html#14) for 3days
  * The role is `REV`
  * The job is not assigned to anyone
  * The [budget](http://www.zerocracy.com/policy.html#4) is not set yet
  * These users are banned and won't be assigned:
    * @paulodamaso[/z](https://www.0crat.com/u/paulodamaso): This user reported the ticket
  * Job [footprint](https://www.0crat.com/footprint/C63314D6Z?q=%7Bjob%3A%27gh%3Ayegor256%2Fcactoos-http%2353%27%2Ctype%3A%7B%24not%3A%2F%5ENotify%2F%7D%7D) (restricted area)

**Comment by 0crat** [comment]
This pull request [#53](https://github.com/yegor256/cactoos-http/issues/53) is assigned to @vssekorin[/z](https://www.0crat.com/u/vssekorin), here is [why](https://www.0crat.com/footprint/C63314D6Z/1361124312); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @llorllale[/z](https://www.0crat.com/u/llorllale) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer

File: src/test/java/org/cactoos/http/HtUpgradeWireTest.java [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+        @Override
+        public InputStream stream() throws IOException {
+            return this.origin.stream();
+        }
+
+        /**
+         * Returns the wire of this request.
+         * @return Wire.
+         */
+        public Wire wire() {
```
Comment (VsSekorin) [inline src/test/java/org/cactoos/http/HtUpgradeWireTest.java]: Public methods without `@Override`. You can `implements Input, Scalar<Wire>`.

File: src/test/java/org/cactoos/http/HtUpgradeWireTest.java [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+    /**
+     * Wrap for response which allows access to its wire.
+     */
+    private class ResponseWrap implements Input {
+
+        /**
+         * Origin response.
+         */
+        private final HtResponse origin;
```
Comment (VsSekorin) [inline src/test/java/org/cactoos/http/HtUpgradeWireTest.java]: Why is this called `origin`, and not a `response`? `origin` is fimilar to decorators, but `HtResponse` is not `Input`.

Comment (paulodamaso) [reply src/test/java/org/cactoos/http/HtUpgradeWireTest.java]: @VsSekorin Didn't unsderstood your remark, `HtResponse` is an `Input` indeed.

Comment (VsSekorin) [reply src/test/java/org/cactoos/http/HtUpgradeWireTest.java]: My bad. Thanks

**Comment by paulodamaso** [comment]
@VsSekorin Implemented suggestions, thank you!

**Comment by VsSekorin** [comment]
@rultor good to merge

**Comment by paulodamaso** [comment]
@llorllale Please take a look.

File: src/main/java/org/cactoos/http/HtUpgradeWire.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+/**
+ * Wire that is capable of upgrading itself upon an 101 status code.
+ * @author Paulo Lobo (pauloeduardolobo@gmail.com)
+ * @version $Id$
+ * @since 0.1
+ * @todo #23:30 min This implementation of class HtUpgradeWire is just
+ *  forwarding origin wire behavior. Finish implementation so a HtUpgradeWire
+ *  can be upgraded to a HtSecureWire on 101 - Swithicng Protocols status code,
+ *  switching the wire to https protocol. The test HtUpgradeWireTest#testUpgrade
```
Comment (llorllale) [inline src/main/java/org/cactoos/http/HtUpgradeWire.java]: @paulodamaso I don't understand why the test needs to be *removed*. Seems like a good outline for a perfectly valid test scenario.

**Comment by llorllale** [comment]
@paulodamaso @VsSekorin not trying to sound harsh, but I am having trouble understanding exactly how this PR moves us forward in the direction of the "upgrade mechanism" of this task, aside from introducing the class' name `HtUpgradeWire` and two tests (one of which apparently we need to remove?)

**Comment by paulodamaso** [comment]
@llorllale Well, this PR does implements the test to guarantee that the 
we will have an `HtSecureWire` when we receive an `101 - Swithcing protocols`. I've faked an 101 code and I am expecting that `HtUpgradeWire` uses an `HtSecureWire`. I've started this PR with the tests of the wire upgrade feature, and it wasn't possible to implement `HtUpgradeWire` class in job timebox; so, I've implemented an dummy implementation of `HtUpgradeWire` just to model my tests, implemented the tests, and left a `@todo` about it.

File: src/main/java/org/cactoos/http/HtUpgradeWire.java (hunk 3 above)
Comment (llorllale) [reply src/main/java/org/cactoos/http/HtUpgradeWire.java]: @paulodamaso please fix the text on this puzzle accordingly

**Comment by llorllale** [comment]
@paulodamaso the outline of the `Upgrade` mechanism: the client initiates by asking for an upgrade in the request (our users will build this request) and if the server responds with the `Upgrade` header then the client needs to support the kind of upgrade specified in the response. Upgrades are typically done from HTTP/1.1 -> HTTP/2, although sometimes HTTP -> HTTPS is used as well.

`HtUpgradeWire` should therefore be configurable; you should be able to specify the wires that you, as the client, are able to upgrade to.

As a rough first draft, we could possibly be do something like:

```java
new HtUpgradeWire(
    new IterableOf<>(
        new MapEntry<Func<String, Boolean>, Func<URI, Wire>>(
            upgrade -> upgrade.contains("TLS"),
            HtSecureWire::new
        )
    )
)
```

**Comment by paulodamaso** [comment]
@llorllale I've corrected `@todo` about implementation of  `HtUpgradeWire`. I think that the above comment should be added to the puzzle that will be created by this `@todo`.

**Comment by llorllale** [comment]
@paulodamaso I'm afraid I can't accept just that. The job of presenting the research rests on the job's assignee. If you had done that then I would have been more willing to accept this PR. 

But now, since I presented the research, I expect more than just a puzzle with the research (unless you do some more research that uncovers *even more* stuff that is relevant to completing the task or that acts as an impediment in some way).

**Comment by paulodamaso** [comment]
@llorllale Please take a look at my comment above, https://github.com/yegor256/cactoos-http/pull/53#issuecomment-391132150 . It explains what I've done in this PR, and why I'm suggesting this course of action. Notice that I'm not just asking a puzzle with the research, I've created the test that models the update wire class behavior according to the documentation provided at the time. I'm afraid that the research presented by you, while useful, is related to puzzle I've added to the code, that asks for `HtUpgradeWire` implementation according to the behavior defined by the test class implemented in this PR.
In my point of view this PR match what was asked in issue description (and trying to follow what said in https://www.yegor256.com/2015/01/15/how-to-cut-corners.html ) ; if you think that I still have to do something here, please clearly define what is it so I can do it and close the issue.

**Comment by llorllale** [comment]
@paulodamaso thinking about it more, the "upgrade" is not the job of a `Wire`, because it would require reading the contents of the response. This is the job of an `HtResponse`-like object. A `Wire` is just a dumb pipe that sends and receives bytes.

So my draft code above should be refactored to:

```java
new HtUpgradedResponse(
    new IterableOf<>(
        new MapEntry<Func<String, Boolean>, Func<URI, Wire>>(
            upgrade -> upgrade.contains("TLS"),
            HtSecureWire::new
        )
    )
)
```

Fix the `@todo` with this and let's call it a day.

**Comment by paulodamaso** [comment]
@llorllale Changes made, take a look.

**Comment by 0crat** [comment]
The job [#53](https://github.com/yegor256/cactoos-http/issues/53) is now out of scope

**Comment by elenavolokhova** [comment]
@VsSekorin According to our Policy:
> The code reviewer found at least three problems in the code

Only two issues were found in this review. Please, confirm that you will try to find at least three problems next time.

**Comment by 0crat** [comment]
Order was finished, quality is "acceptable": +15 point(s) just awarded to @vssekorin[/z](https://www.0crat.com/u/vssekorin)

**Comment by 0crat** [comment]
Quality review completed: +8 point(s) just awarded to @elenavolokhova[/z](https://www.0crat.com/u/elenavolokhova)

**Pull request merged**